File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.cpp
#include <bits/stdc++.h>
using namespace std;
typedef vector<vector<int>> vi;
const int N = 502;
int n;
int dp[N][N];

int solve(int x, int y, vi &a)
{
    int &ret = dp[x][y];
    if (ret != -1) return ret;
    ret = a[x][y];
    if (x == n - 1) return ret;
    int res = max(solve(x + 1, y, a), solve(x + 1, y + 1, a));
    ret += res;
    return ret;
}

int solution(vi a) 
{
    n = a.size();
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j <= i; j++)
            dp[i][j] = -1;
    }
    return solve(0, 0, a);
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
import java.util.*;

class Solution 
{
    static final int N = 502;
    int[][] a = new int[N][N];
    int[][] dp = new int[N][N];
    int n, m;
    
    public int solve(int x, int y)
    {
        if (dp[x][y] != -1) return dp[x][y];
        if (x == n) return dp[x][y] = 0;
        dp[x][y] = a[x][y];
        int res = Math.max(solve(x + 1, y), solve(x + 1, y + 1));
        dp[x][y] += res;
        return dp[x][y];
    }
    
    public int solution(int[][] v) 
    {
        n = v.length;
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < v[i].length; j++)
            {
                a[i][j] = v[i][j];
                dp[i][j] = -1;
            }
        }
        return solve(0, 0);
    }
}